### ISSUE TITLE
Descrepancy in lattice optimization between DFTB+ and CP2K

### ISSUE THREAD

**Issue by xavbdlt** [dftbplus/dftbplus#1001]
Hello,

I run a variable-cell optimization of a high-energy molecular crystal, beta-HMX, with CP2K 9.1 and DFTB+ 21.2, both using GFN1-xTB, which includes D3(BJ) correction. To compare apples to apples, I optimized the same supercell, using the Gamma point. The initial (experimental) volume for the supercell is 2069.80 A3. The resulting volume is:
1950.83 A3 with DFTB+
1759.92 A3 with CP2K

In order to check if I did something wrong in the configuration files, I used DFTB + D3(0) with the default DFTB parameter file from CP2K for both CP2K and DFTB+, and I find:
1467.79 A3 with DFTB+
1467.14 A3 with CP2K
So, DFTB + D3(0) is consistent between CP2K and DFTB+, but not GFN1-xTB.

My 1st guess is that the dispersion correction of GFN1-xTB is wrong somewhere, either in DFTB+ or in CP2K. Looking at the difference between CP2K and DFTB+, could it be that D3(BJ) correction is counted twice in CP2K? Or with some wrong units we don't have access to?

I attached the DFTB+ input file for the GFN1-xTB calculation.
[dftb_in.hsd.zip](https://github.com/dftbplus/dftbplus/files/8447092/dftb_in.hsd.zip)

I also posted in CP2K group (https://groups.google.com/g/cp2k/c/lESdjFfgq3g/m/COZ3Ii4MCQAJ ) with the input script for CP2K.

Please, let me know.

Thank you,
Xavier

**Comment by awvwgk** [comment]
First, the `ConjugateGradient` driver in DFTB+ is not capable of correctly optimizing lattice parameters as it cannot reach the actual minimum structure. I implemented a new optimization driver without this issues (`GeometryOptimization`), which should be able to perform a proper lattice relaxation.

Second, I recall their being some discrepancies between the CP2K and `xtb` basis set setup, especially in the choice of primitives. If you try a molecular calculation with CP2K, DFTB+ and `xtb` I would presume that there will already be a mismatch in this case between CP2K and DFTB+/`xtb`.

**Comment by xavbdlt** [comment]
Thank you for your feedback. For improved comparison, I additionally used CG with CP2K, and various driver/optimizer couples for DFTB+. Here are the results for the volume and total energy:

CP2K:
BFGS -> 1759.92 A3 and -16435.8400 eV
CG -> 1759.11 A3 and -16435.8394 eV
That's why I assumed CG would work great with DFTB+ as well, but apparently not. Note that deactivating the symmetry constraint yields similar results, and with well-preserved 90° angles (relative error less than 1e-7).

DFTB+:
CG -> 1950.83 A3 and -16434.7164 eV
LBFGS -> 1950.77 A3 and -16434.7164 eV
GO/Rational -> 1859.16 A3 and -16435.2202 eV (input file [dftb_in.hsd.zip](https://github.com/dftbplus/dftbplus/files/8454915/dftb_in.hsd.zip))
GO/LBFGS -> 1906.10 A3 and -16434.8907 eV (on this one, the 90° angles are not well-preserved, modifying even the symmetry group upon analysis of the resulting cif file with Avogadro)

So, for DFTB+, the drivers CG and LBFGS yield very similar converged values, which are not the lowest energy though. The couple GO/LBFGS seems not so reliable as well. It seems the best couple is GO/rational, which gives the lowest energy for DFTB+, but there is still ~100 A3 difference between CP2K and DFTB+. No typo (I double checked the values).

From 2nd point, this difference would come from the different basis sets between CP2K and DFTB+/xtb. Could you elaborate?

Thank you,
Xavier

**Comment by awvwgk** [comment]
You are comparing two different things here, the efficiency of the geometry optimization driver in CP2K and DFTB+ and the actual implementation of an electronic structure method, namely GFN1-xTB.

For the first comparison, make sure that you actually select the same convergence thresholds for CP2K and DFTB+, using the defaults for both programs is unlikely to be comparable between the two programs. I highly recommend to stick with the GO drivers, as the CG and LBFGS drivers are actual erroneous for lattice optimizations, hence the bad high resulting energy.

For the latter comparison, best start with performing just a single point calculation on the identical input structure with DFTB+ and CP2K, otherwise you will mix two different effects in your test. You might find a slight discrepancy between CP2K and DFTB+ because of the choice of basis set in CP2K for xTB, *i.e.* using the same contraction length for all STO-NG basis functions, while in DFTB+ the contraction length is specific for each element.

**Comment by xavbdlt** [comment]
I tried first an initial single calculation, and from it a restart for the geometry optimization. I've got very similar results. So, starting straight from the experimental geometry replicated as a 2x1x2 initial box works great.

I've tested convergence parameters of 1e-5 or 1e-6 for geometry convergence, and again, no significant difference.

I sticked to GO/rational for DFTB+ and explored a few SCC epsilon parameters (expressed below as 1e-n), as well for CP2K (SCF epsilon). 

CP2K (all other epsilons are default):
n   Volume(A3)
5   1870.51
6   1863.46
7   1873.30
8   1759.92
9   1834.73
10 1944.21
11 1761.18
12 1846.56
13 1860.54
14 1826.25
15 1852.36

The differences are large, and non-monotonic. There is, imo, no obvious choice. I also quickly checked that without constraining the monoclinic geometry, it was actually preserved.

DFTB+:
5   1901.99
6   1859.22
7   1904.44
8   1902.72
9   1900.05
10 1900.89
11 1901.01
12 1900.90
13 1900.35

Note that the only one preserving the monoclinic geometry is 1e-6. Alpha and Gamma are around 90.2° and 90.7°, respectively, for all the other epsilons. But all those other volumes are very similar, and consistent with GO/lbfgs of the previous message. Would it mean we have to avoid 1e-6 for  DFTB+? Or the contrary?

Any other remarks that could help?

Thank you,
Xavier

**Comment by awvwgk** [comment]
> I've got very similar results.

That's good to hear, so the GFN1-xTB implementation in CP2K is consistent.

**Comment by xavbdlt** [comment]
Looking back at the output structures obtained from DFTB+, I just noticed that the one that didn't mess with 90° alpha and gamma angles (the one using SCC epsilon of 1e-6) is also the only yielding the "b" dimension under 10 Angströms. (9.64 A). All the other yielded "b" slighly above 10 Ansgtröms (around 10.04 A). Might this be something related to a cutoff somewhere?

**Comment by xavbdlt** [comment]
Good news. I tried again yesterday but with dftb+ 22.1. I don't know what has changed, but the results are now consistent across the electronic convergence parameters.

DFTB+ 22.1:
Epsilon(1e-n) Volume (A3)
5 1858.44
6 1858.41 
7 1858.43
8 1858.45
9 1858.44
10 1858.45
11 1858.40
12 1858.40

And now, the 90° angles are well-preserved.

Good with OpenMP (2 threads) or MPI (1 or 2 threads), with DIIS or Broyden. I don't know what has changed but everything is good now. I attached the input script.
[dftb_in.hsd.zip](https://github.com/dftbplus/dftbplus/files/9151240/dftb_in.hsd.zip)
.
